ivadomed/loader/loader.py

```diff
@@ -817,104 +825,94 @@ def __init__(self, path_data, subject_lst, target_suffix, contrast_params, slice
 
-                    metadata['metadata_dict'] = metadata_dict
-
-                # Fill multichannel dictionary
-                if multichannel:
-                    idx = idx_dict[subject.record["modality"]]
-                    subj_id = subject.record["subject_id"]
-                    multichannel_subjects[subj_id]["absolute_paths"][idx] = subject.record.absolute_path
-                    multichannel_subjects[subj_id]["deriv_path"] = target_filename
-                    multichannel_subjects[subj_id]["metadata"][idx] = metadata
-                    if roi_filename:
-                        multichannel_subjects[subj_id]["roi_filename"] = roi_filename
+            # Training & Validation: do not consider the contrasts over the threshold contained in contrast_balance
+            contrast = df_sub['suffix'].values[0]
+            if contrast in (contrast_params["balance"].keys()):
+                c[contrast] = c[contrast] + 1
+                if c[contrast] / tot[contrast] > contrast_params["balance"][contrast]:
+                    continue
+            if isinstance(target_suffix[0], str):
+                target_filename, roi_filename = [None] * len(target_suffix), None
+            else:
+                target_filename, roi_filename = [[] for _ in range(len(target_suffix))], None
+
+            derivatives = bids_df.df[bids_df.df['filename']
+                          .str.contains('|'.join(bids_df.get_derivatives(subject, all_deriv)))]['path'].to_list()
+
+            for deriv in derivatives:
+                for idx, suffix_list in enumerate(target_suffix):
+                    # If suffix_list is a string, then only one rater annotation per class is available.
+                    # Otherwise, multiple raters segmented the same class.
+                    if isinstance(suffix_list, list):
+                        for suffix in suffix_list:
+                            if suffix in deriv:
+                                target_filename[idx].append(deriv)
+                    elif suffix_list in deriv:
+                        target_filename[idx] = deriv
+                if not (self.roi_params["suffix"] is None) and self.roi_params["suffix"] in deriv:
+                    roi_filename = [deriv]
+
+            if (not any(target_filename)) or (not (self.roi_params["suffix"] is None) and (roi_filename is None)):
+                continue
+
+            metadata = df_sub.to_dict(orient='records')[0]
+            metadata['contrast'] = contrast
+
+            if len(bounding_box_dict):
+                # Take only one bounding box for cropping
+                metadata['bounding_box'] = bounding_box_dict[str(df_sub['path'].values[0])][0]
+
+            if metadata_choice == 'mri_params':
+                if not all([imed_film.check_isMRIparam(m, metadata, subject, self.metadata) for m in
+                            self.metadata.keys()]):
+                    continue
```

Comment:
there is a lot of code nested within the for loop from line 844. Perhaps there is a way to extract some of the logic into other functions for better readability

Reply:
I agree with you. However I think this is out of scope for this already big PR, I suggest we open a separate issue for this.